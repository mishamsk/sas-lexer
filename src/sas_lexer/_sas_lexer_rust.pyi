def lex_str(src: str) -> bytes: ...
