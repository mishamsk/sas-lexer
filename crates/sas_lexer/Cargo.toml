[package]
name = "sas_lexer"
authors = { workspace = true }
edition = { workspace = true }
publish = { workspace = true }
repository = { workspace = true }
rust-version = { workspace = true }
version = { workspace = true }

[lib]

[dependencies]
phf = { workspace = true }
sas_lexer_macro = { workspace = true }
rmp-serde = { workspace = true, optional = true }
serde = { workspace = true, features = ["derive"], optional = true }
serde_repr = { version = "0.1.19", optional = true }
strum = { workspace = true }
unicode-ident = "1.0.12"
lexical = { version = "7.0.2", default-features = false, features = ["std", "power-of-two", "parse-floats", "parse-integers", "format"] }

[dev-dependencies]
criterion = { workspace = true, features = ["html_reports"] }
insta = { workspace = true }
rstest = { workspace = true }
zip = { workspace = true }
walkdir = { workspace = true }
pprof = { workspace = true, features = ["flamegraph", "criterion"] }

[features]
serde = ["dep:serde", "dep:rmp-serde", "dep:serde_repr"]

[lints]
workspace = true

[[bench]]
name = "lexer"
harness = false
