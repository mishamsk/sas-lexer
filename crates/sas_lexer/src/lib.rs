mod lexer;

pub use lexer::buffer::{Payload, ResolvedTokenInfo, TokenIdx, TokenizedBuffer};
pub use lexer::channel::TokenChannel;
pub use lexer::error;
pub use lexer::lex;
pub use lexer::token_type::TokenType;
