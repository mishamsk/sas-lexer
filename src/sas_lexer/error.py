from msgspec import Struct


class Error(Struct, array_like=True, gc=False, frozen=True):
    error_kind: int
    at_byte_offset: int
    at_char_offset: int
    on_line: int
    at_column: int
    last_token_index: int | None
