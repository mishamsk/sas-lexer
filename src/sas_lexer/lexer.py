from collections.abc import Sequence

from msgspec.msgpack import Decoder

from sas_lexer._sas_lexer_rust import lex_program_from_str as _lex_program_from_str
from sas_lexer.token import Token

TOK_LIST_DECODER = Decoder(list[Token])


def lex_program_from_str(source: str) -> Sequence[Token]:
    return TOK_LIST_DECODER.decode(_lex_program_from_str(source))
