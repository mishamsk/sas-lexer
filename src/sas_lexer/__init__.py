from sas_lexer.lexer import lex_program_from_str

__all__ = ["lex_program_from_str"]
