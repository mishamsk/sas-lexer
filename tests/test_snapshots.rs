#[macro_use]
mod util;

use insta::assert_yaml_snapshot;
use rstest::rstest;
use sas_lexer::{lex, MaterializedToken};
use std::{fs, path::PathBuf};

#[rstest]
fn test_files(#[files("tests/samples/**/*.sas")] path: PathBuf) {
    // Compute the absolute path of the prefix
    let prefix = fs::canonicalize("tests/samples/").unwrap();

    let snap_name = path.strip_prefix(&prefix).unwrap();
    let snap_name_str = snap_name
        .to_str()
        .unwrap()
        .replace(std::path::MAIN_SEPARATOR, "__");

    set_snapshot_suffix!("{}", snap_name_str);

    let contents = fs::read_to_string(&path).unwrap();
    let tok_buffer = lex(contents.as_str());
    let tokens: Vec<MaterializedToken> = tok_buffer
        .into_iter()
        .map(|tidx| tok_buffer.materialize_token(tidx))
        .collect();

    assert_yaml_snapshot!(tokens);
}
