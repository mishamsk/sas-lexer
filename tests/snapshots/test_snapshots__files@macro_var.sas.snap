---
source: tests/test_snapshots.rs
expression: tokens
---
- "[@0,0:1=\"&\",<BaseCode>,L1:C0-L1:C1,chl=<D>,pl=None]"
- "[@1,1:2=\"\\n\",<WS>,L1:C1-L1:C2,chl=<H>,pl=None]"
- "[@2,2:3=\"&\",<BaseCode>,L2:C0-L2:C1,chl=<D>,pl=None]"
- "[@3,3:4=\"9\",<BaseCode>,L2:C1-L2:C2,chl=<D>,pl=None]"
- "[@4,4:5=\"\\n\",<WS>,L2:C2-L2:C3,chl=<H>,pl=None]"
- "[@5,5:6=\"&\",<BaseCode>,L3:C0-L3:C1,chl=<D>,pl=None]"
- "[@6,6:7=\"m\",<BaseCode>,L3:C1-L3:C2,chl=<D>,pl=None]"
- "[@7,7:8=\"v\",<BaseCode>,L3:C2-L3:C3,chl=<D>,pl=None]"
- "[@8,8:9=\"a\",<BaseCode>,L3:C3-L3:C4,chl=<D>,pl=None]"
- "[@9,9:10=\"r\",<BaseCode>,L3:C4-L3:C5,chl=<D>,pl=None]"
- "[@10,10:11=\"\\n\",<WS>,L3:C5-L3:C6,chl=<H>,pl=None]"
- "[@11,11:12=\"&\",<BaseCode>,L4:C0-L4:C1,chl=<D>,pl=None]"
- "[@12,12:13=\"m\",<BaseCode>,L4:C1-L4:C2,chl=<D>,pl=None]"
- "[@13,13:14=\"v\",<BaseCode>,L4:C2-L4:C3,chl=<D>,pl=None]"
- "[@14,14:15=\"a\",<BaseCode>,L4:C3-L4:C4,chl=<D>,pl=None]"
- "[@15,15:16=\"r\",<BaseCode>,L4:C4-L4:C5,chl=<D>,pl=None]"
- "[@16,16:17=\".\",<BaseCode>,L4:C5-L4:C6,chl=<D>,pl=None]"
- "[@17,17:18=\"\\n\",<WS>,L4:C6-L4:C7,chl=<H>,pl=None]"
- "[@18,18:19=\"&\",<BaseCode>,L5:C0-L5:C1,chl=<D>,pl=None]"
- "[@19,19:20=\"m\",<BaseCode>,L5:C1-L5:C2,chl=<D>,pl=None]"
- "[@20,20:21=\"v\",<BaseCode>,L5:C2-L5:C3,chl=<D>,pl=None]"
- "[@21,21:22=\"a\",<BaseCode>,L5:C3-L5:C4,chl=<D>,pl=None]"
- "[@22,22:23=\"r\",<BaseCode>,L5:C4-L5:C5,chl=<D>,pl=None]"
- "[@23,23:24=\"&\",<BaseCode>,L5:C5-L5:C6,chl=<D>,pl=None]"
- "[@24,24:25=\"\\n\",<WS>,L5:C6-L5:C7,chl=<H>,pl=None]"
- "[@25,25:26=\"&\",<BaseCode>,L6:C0-L6:C1,chl=<D>,pl=None]"
- "[@26,26:27=\"m\",<BaseCode>,L6:C1-L6:C2,chl=<D>,pl=None]"
- "[@27,27:28=\"v\",<BaseCode>,L6:C2-L6:C3,chl=<D>,pl=None]"
- "[@28,28:29=\"a\",<BaseCode>,L6:C3-L6:C4,chl=<D>,pl=None]"
- "[@29,29:30=\"r\",<BaseCode>,L6:C4-L6:C5,chl=<D>,pl=None]"
- "[@30,30:31=\".\",<BaseCode>,L6:C5-L6:C6,chl=<D>,pl=None]"
- "[@31,31:32=\"&\",<BaseCode>,L6:C6-L6:C7,chl=<D>,pl=None]"
- "[@32,32:33=\"\\n\",<WS>,L6:C7-L6:C8,chl=<H>,pl=None]"
- "[@33,33:34=\"&\",<BaseCode>,L7:C0-L7:C1,chl=<D>,pl=None]"
- "[@34,34:35=\"m\",<BaseCode>,L7:C1-L7:C2,chl=<D>,pl=None]"
- "[@35,35:36=\"v\",<BaseCode>,L7:C2-L7:C3,chl=<D>,pl=None]"
- "[@36,36:37=\"a\",<BaseCode>,L7:C3-L7:C4,chl=<D>,pl=None]"
- "[@37,37:38=\"r\",<BaseCode>,L7:C4-L7:C5,chl=<D>,pl=None]"
- "[@38,38:39=\"&\",<BaseCode>,L7:C5-L7:C6,chl=<D>,pl=None]"
- "[@39,39:40=\".\",<BaseCode>,L7:C6-L7:C7,chl=<D>,pl=None]"
- "[@40,40:41=\"\\n\",<WS>,L7:C7-L7:C8,chl=<H>,pl=None]"
- "[@41,41:42=\"&\",<BaseCode>,L8:C0-L8:C1,chl=<D>,pl=None]"
- "[@42,42:43=\"&\",<BaseCode>,L8:C1-L8:C2,chl=<D>,pl=None]"
- "[@43,43:44=\"m\",<BaseCode>,L8:C2-L8:C3,chl=<D>,pl=None]"
- "[@44,44:45=\"v\",<BaseCode>,L8:C3-L8:C4,chl=<D>,pl=None]"
- "[@45,45:46=\"a\",<BaseCode>,L8:C4-L8:C5,chl=<D>,pl=None]"
- "[@46,46:47=\"r\",<BaseCode>,L8:C5-L8:C6,chl=<D>,pl=None]"
- "[@47,47:48=\"\\n\",<WS>,L8:C6-L8:C7,chl=<H>,pl=None]"
- "[@48,48:49=\"&\",<BaseCode>,L9:C0-L9:C1,chl=<D>,pl=None]"
- "[@49,49:50=\"&\",<BaseCode>,L9:C1-L9:C2,chl=<D>,pl=None]"
- "[@50,50:51=\"m\",<BaseCode>,L9:C2-L9:C3,chl=<D>,pl=None]"
- "[@51,51:52=\"v\",<BaseCode>,L9:C3-L9:C4,chl=<D>,pl=None]"
- "[@52,52:53=\"a\",<BaseCode>,L9:C4-L9:C5,chl=<D>,pl=None]"
- "[@53,53:54=\"r\",<BaseCode>,L9:C5-L9:C6,chl=<D>,pl=None]"
- "[@54,54:55=\".\",<BaseCode>,L9:C6-L9:C7,chl=<D>,pl=None]"
- "[@55,55:56=\"\\n\",<WS>,L9:C7-L9:C8,chl=<H>,pl=None]"
- "[@56,56:57=\"&\",<BaseCode>,L10:C0-L10:C1,chl=<D>,pl=None]"
- "[@57,57:58=\"&\",<BaseCode>,L10:C1-L10:C2,chl=<D>,pl=None]"
- "[@58,58:59=\"m\",<BaseCode>,L10:C2-L10:C3,chl=<D>,pl=None]"
- "[@59,59:60=\"v\",<BaseCode>,L10:C3-L10:C4,chl=<D>,pl=None]"
- "[@60,60:61=\"a\",<BaseCode>,L10:C4-L10:C5,chl=<D>,pl=None]"
- "[@61,61:62=\"r\",<BaseCode>,L10:C5-L10:C6,chl=<D>,pl=None]"
- "[@62,62:63=\"&\",<BaseCode>,L10:C6-L10:C7,chl=<D>,pl=None]"
- "[@63,63:64=\"\\n\",<WS>,L10:C7-L10:C8,chl=<H>,pl=None]"
- "[@64,64:65=\"&\",<BaseCode>,L11:C0-L11:C1,chl=<D>,pl=None]"
- "[@65,65:66=\"&\",<BaseCode>,L11:C1-L11:C2,chl=<D>,pl=None]"
- "[@66,66:67=\"m\",<BaseCode>,L11:C2-L11:C3,chl=<D>,pl=None]"
- "[@67,67:68=\"v\",<BaseCode>,L11:C3-L11:C4,chl=<D>,pl=None]"
- "[@68,68:69=\"a\",<BaseCode>,L11:C4-L11:C5,chl=<D>,pl=None]"
- "[@69,69:70=\"r\",<BaseCode>,L11:C5-L11:C6,chl=<D>,pl=None]"
- "[@70,70:71=\".\",<BaseCode>,L11:C6-L11:C7,chl=<D>,pl=None]"
- "[@71,71:72=\"&\",<BaseCode>,L11:C7-L11:C8,chl=<D>,pl=None]"
- "[@72,72:73=\"\\n\",<WS>,L11:C8-L11:C9,chl=<H>,pl=None]"
- "[@73,73:74=\"&\",<BaseCode>,L12:C0-L12:C1,chl=<D>,pl=None]"
- "[@74,74:75=\"&\",<BaseCode>,L12:C1-L12:C2,chl=<D>,pl=None]"
- "[@75,75:76=\"m\",<BaseCode>,L12:C2-L12:C3,chl=<D>,pl=None]"
- "[@76,76:77=\"v\",<BaseCode>,L12:C3-L12:C4,chl=<D>,pl=None]"
- "[@77,77:78=\"a\",<BaseCode>,L12:C4-L12:C5,chl=<D>,pl=None]"
- "[@78,78:79=\"r\",<BaseCode>,L12:C5-L12:C6,chl=<D>,pl=None]"
- "[@79,79:80=\"&\",<BaseCode>,L12:C6-L12:C7,chl=<D>,pl=None]"
- "[@80,80:81=\"9\",<BaseCode>,L12:C7-L12:C8,chl=<D>,pl=None]"
- "[@81,81:82=\"\\n\",<WS>,L12:C8-L12:C9,chl=<H>,pl=None]"
- "[@82,82:83=\"&\",<BaseCode>,L13:C0-L13:C1,chl=<D>,pl=None]"
- "[@83,83:84=\"&\",<BaseCode>,L13:C1-L13:C2,chl=<D>,pl=None]"
- "[@84,84:85=\"m\",<BaseCode>,L13:C2-L13:C3,chl=<D>,pl=None]"
- "[@85,85:86=\"v\",<BaseCode>,L13:C3-L13:C4,chl=<D>,pl=None]"
- "[@86,86:87=\"a\",<BaseCode>,L13:C4-L13:C5,chl=<D>,pl=None]"
- "[@87,87:88=\"r\",<BaseCode>,L13:C5-L13:C6,chl=<D>,pl=None]"
- "[@88,88:89=\".\",<BaseCode>,L13:C6-L13:C7,chl=<D>,pl=None]"
- "[@89,89:90=\"&\",<BaseCode>,L13:C7-L13:C8,chl=<D>,pl=None]"
- "[@90,90:91=\"9\",<BaseCode>,L13:C8-L13:C9,chl=<D>,pl=None]"
- "[@91,91:92=\"\\n\",<WS>,L13:C9-L13:C10,chl=<H>,pl=None]"
- "[@92,92:93=\"&\",<BaseCode>,L14:C0-L14:C1,chl=<D>,pl=None]"
- "[@93,93:94=\"m\",<BaseCode>,L14:C1-L14:C2,chl=<D>,pl=None]"
- "[@94,94:95=\"v\",<BaseCode>,L14:C2-L14:C3,chl=<D>,pl=None]"
- "[@95,95:96=\"a\",<BaseCode>,L14:C3-L14:C4,chl=<D>,pl=None]"
- "[@96,96:97=\"r\",<BaseCode>,L14:C4-L14:C5,chl=<D>,pl=None]"
- "[@97,97:98=\"&\",<BaseCode>,L14:C5-L14:C6,chl=<D>,pl=None]"
- "[@98,98:99=\"m\",<BaseCode>,L14:C6-L14:C7,chl=<D>,pl=None]"
- "[@99,99:100=\"v\",<BaseCode>,L14:C7-L14:C8,chl=<D>,pl=None]"
- "[@100,100:101=\"a\",<BaseCode>,L14:C8-L14:C9,chl=<D>,pl=None]"
- "[@101,101:102=\"r\",<BaseCode>,L14:C9-L14:C10,chl=<D>,pl=None]"
- "[@102,102:103=\"2\",<BaseCode>,L14:C10-L14:C11,chl=<D>,pl=None]"
- "[@103,103:104=\"\\n\",<WS>,L14:C11-L14:C12,chl=<H>,pl=None]"
- "[@104,104:105=\"&\",<BaseCode>,L15:C0-L15:C1,chl=<D>,pl=None]"
- "[@105,105:106=\"m\",<BaseCode>,L15:C1-L15:C2,chl=<D>,pl=None]"
- "[@106,106:107=\"v\",<BaseCode>,L15:C2-L15:C3,chl=<D>,pl=None]"
- "[@107,107:108=\"a\",<BaseCode>,L15:C3-L15:C4,chl=<D>,pl=None]"
- "[@108,108:109=\"r\",<BaseCode>,L15:C4-L15:C5,chl=<D>,pl=None]"
- "[@109,109:110=\"&\",<BaseCode>,L15:C5-L15:C6,chl=<D>,pl=None]"
- "[@110,110:111=\"m\",<BaseCode>,L15:C6-L15:C7,chl=<D>,pl=None]"
- "[@111,111:112=\"v\",<BaseCode>,L15:C7-L15:C8,chl=<D>,pl=None]"
- "[@112,112:113=\"a\",<BaseCode>,L15:C8-L15:C9,chl=<D>,pl=None]"
- "[@113,113:114=\"r\",<BaseCode>,L15:C9-L15:C10,chl=<D>,pl=None]"
- "[@114,114:115=\"2\",<BaseCode>,L15:C10-L15:C11,chl=<D>,pl=None]"
- "[@115,115:116=\".\",<BaseCode>,L15:C11-L15:C12,chl=<D>,pl=None]"
- "[@116,116:117=\"\\n\",<WS>,L15:C12-L15:C13,chl=<H>,pl=None]"
- "[@117,117:118=\"&\",<BaseCode>,L16:C0-L16:C1,chl=<D>,pl=None]"
- "[@118,118:119=\"m\",<BaseCode>,L16:C1-L16:C2,chl=<D>,pl=None]"
- "[@119,119:120=\"v\",<BaseCode>,L16:C2-L16:C3,chl=<D>,pl=None]"
- "[@120,120:121=\"a\",<BaseCode>,L16:C3-L16:C4,chl=<D>,pl=None]"
- "[@121,121:122=\"r\",<BaseCode>,L16:C4-L16:C5,chl=<D>,pl=None]"
- "[@122,122:123=\".\",<BaseCode>,L16:C5-L16:C6,chl=<D>,pl=None]"
- "[@123,123:124=\"&\",<BaseCode>,L16:C6-L16:C7,chl=<D>,pl=None]"
- "[@124,124:125=\"m\",<BaseCode>,L16:C7-L16:C8,chl=<D>,pl=None]"
- "[@125,125:126=\"v\",<BaseCode>,L16:C8-L16:C9,chl=<D>,pl=None]"
- "[@126,126:127=\"a\",<BaseCode>,L16:C9-L16:C10,chl=<D>,pl=None]"
- "[@127,127:128=\"r\",<BaseCode>,L16:C10-L16:C11,chl=<D>,pl=None]"
- "[@128,128:129=\"2\",<BaseCode>,L16:C11-L16:C12,chl=<D>,pl=None]"
- "[@129,129:130=\"\\n\",<WS>,L16:C12-L16:C13,chl=<H>,pl=None]"
- "[@130,130:131=\"&\",<BaseCode>,L17:C0-L17:C1,chl=<D>,pl=None]"
- "[@131,131:132=\"m\",<BaseCode>,L17:C1-L17:C2,chl=<D>,pl=None]"
- "[@132,132:133=\"v\",<BaseCode>,L17:C2-L17:C3,chl=<D>,pl=None]"
- "[@133,133:134=\"a\",<BaseCode>,L17:C3-L17:C4,chl=<D>,pl=None]"
- "[@134,134:135=\"r\",<BaseCode>,L17:C4-L17:C5,chl=<D>,pl=None]"
- "[@135,135:136=\".\",<BaseCode>,L17:C5-L17:C6,chl=<D>,pl=None]"
- "[@136,136:137=\"&\",<BaseCode>,L17:C6-L17:C7,chl=<D>,pl=None]"
- "[@137,137:138=\"m\",<BaseCode>,L17:C7-L17:C8,chl=<D>,pl=None]"
- "[@138,138:139=\"v\",<BaseCode>,L17:C8-L17:C9,chl=<D>,pl=None]"
- "[@139,139:140=\"a\",<BaseCode>,L17:C9-L17:C10,chl=<D>,pl=None]"
- "[@140,140:141=\"r\",<BaseCode>,L17:C10-L17:C11,chl=<D>,pl=None]"
- "[@141,141:142=\"2\",<BaseCode>,L17:C11-L17:C12,chl=<D>,pl=None]"
- "[@142,142:143=\".\",<BaseCode>,L17:C12-L17:C13,chl=<D>,pl=None]"
- "[@143,143:144=\"\\n\",<WS>,L17:C13-L17:C14,chl=<H>,pl=None]"
- "[@144,144:145=\"&\",<BaseCode>,L18:C0-L18:C1,chl=<D>,pl=None]"
- "[@145,145:146=\"&\",<BaseCode>,L18:C1-L18:C2,chl=<D>,pl=None]"
- "[@146,146:147=\"m\",<BaseCode>,L18:C2-L18:C3,chl=<D>,pl=None]"
- "[@147,147:148=\"v\",<BaseCode>,L18:C3-L18:C4,chl=<D>,pl=None]"
- "[@148,148:149=\"a\",<BaseCode>,L18:C4-L18:C5,chl=<D>,pl=None]"
- "[@149,149:150=\"r\",<BaseCode>,L18:C5-L18:C6,chl=<D>,pl=None]"
- "[@150,150:151=\"&\",<BaseCode>,L18:C6-L18:C7,chl=<D>,pl=None]"
- "[@151,151:152=\"m\",<BaseCode>,L18:C7-L18:C8,chl=<D>,pl=None]"
- "[@152,152:153=\"v\",<BaseCode>,L18:C8-L18:C9,chl=<D>,pl=None]"
- "[@153,153:154=\"a\",<BaseCode>,L18:C9-L18:C10,chl=<D>,pl=None]"
- "[@154,154:155=\"r\",<BaseCode>,L18:C10-L18:C11,chl=<D>,pl=None]"
- "[@155,155:156=\"2\",<BaseCode>,L18:C11-L18:C12,chl=<D>,pl=None]"
- "[@156,156:157=\"\\n\",<WS>,L18:C12-L18:C13,chl=<H>,pl=None]"
- "[@157,157:158=\"&\",<BaseCode>,L19:C0-L19:C1,chl=<D>,pl=None]"
- "[@158,158:159=\"&\",<BaseCode>,L19:C1-L19:C2,chl=<D>,pl=None]"
- "[@159,159:160=\"m\",<BaseCode>,L19:C2-L19:C3,chl=<D>,pl=None]"
- "[@160,160:161=\"v\",<BaseCode>,L19:C3-L19:C4,chl=<D>,pl=None]"
- "[@161,161:162=\"a\",<BaseCode>,L19:C4-L19:C5,chl=<D>,pl=None]"
- "[@162,162:163=\"r\",<BaseCode>,L19:C5-L19:C6,chl=<D>,pl=None]"
- "[@163,163:164=\"&\",<BaseCode>,L19:C6-L19:C7,chl=<D>,pl=None]"
- "[@164,164:165=\"m\",<BaseCode>,L19:C7-L19:C8,chl=<D>,pl=None]"
- "[@165,165:166=\"v\",<BaseCode>,L19:C8-L19:C9,chl=<D>,pl=None]"
- "[@166,166:167=\"a\",<BaseCode>,L19:C9-L19:C10,chl=<D>,pl=None]"
- "[@167,167:168=\"r\",<BaseCode>,L19:C10-L19:C11,chl=<D>,pl=None]"
- "[@168,168:169=\"2\",<BaseCode>,L19:C11-L19:C12,chl=<D>,pl=None]"
- "[@169,169:170=\".\",<BaseCode>,L19:C12-L19:C13,chl=<D>,pl=None]"
- "[@170,170:171=\"\\n\",<WS>,L19:C13-L19:C14,chl=<H>,pl=None]"
- "[@171,171:172=\"&\",<BaseCode>,L20:C0-L20:C1,chl=<D>,pl=None]"
- "[@172,172:173=\"&\",<BaseCode>,L20:C1-L20:C2,chl=<D>,pl=None]"
- "[@173,173:174=\"m\",<BaseCode>,L20:C2-L20:C3,chl=<D>,pl=None]"
- "[@174,174:175=\"v\",<BaseCode>,L20:C3-L20:C4,chl=<D>,pl=None]"
- "[@175,175:176=\"a\",<BaseCode>,L20:C4-L20:C5,chl=<D>,pl=None]"
- "[@176,176:177=\"r\",<BaseCode>,L20:C5-L20:C6,chl=<D>,pl=None]"
- "[@177,177:178=\".\",<BaseCode>,L20:C6-L20:C7,chl=<D>,pl=None]"
- "[@178,178:179=\"&\",<BaseCode>,L20:C7-L20:C8,chl=<D>,pl=None]"
- "[@179,179:180=\"m\",<BaseCode>,L20:C8-L20:C9,chl=<D>,pl=None]"
- "[@180,180:181=\"v\",<BaseCode>,L20:C9-L20:C10,chl=<D>,pl=None]"
- "[@181,181:182=\"a\",<BaseCode>,L20:C10-L20:C11,chl=<D>,pl=None]"
- "[@182,182:183=\"r\",<BaseCode>,L20:C11-L20:C12,chl=<D>,pl=None]"
- "[@183,183:184=\"2\",<BaseCode>,L20:C12-L20:C13,chl=<D>,pl=None]"
- "[@184,184:185=\"\\n\",<WS>,L20:C13-L20:C14,chl=<H>,pl=None]"
- "[@185,185:186=\"&\",<BaseCode>,L21:C0-L21:C1,chl=<D>,pl=None]"
- "[@186,186:187=\"&\",<BaseCode>,L21:C1-L21:C2,chl=<D>,pl=None]"
- "[@187,187:188=\"m\",<BaseCode>,L21:C2-L21:C3,chl=<D>,pl=None]"
- "[@188,188:189=\"v\",<BaseCode>,L21:C3-L21:C4,chl=<D>,pl=None]"
- "[@189,189:190=\"a\",<BaseCode>,L21:C4-L21:C5,chl=<D>,pl=None]"
- "[@190,190:191=\"r\",<BaseCode>,L21:C5-L21:C6,chl=<D>,pl=None]"
- "[@191,191:192=\".\",<BaseCode>,L21:C6-L21:C7,chl=<D>,pl=None]"
- "[@192,192:193=\"&\",<BaseCode>,L21:C7-L21:C8,chl=<D>,pl=None]"
- "[@193,193:194=\"m\",<BaseCode>,L21:C8-L21:C9,chl=<D>,pl=None]"
- "[@194,194:195=\"v\",<BaseCode>,L21:C9-L21:C10,chl=<D>,pl=None]"
- "[@195,195:196=\"a\",<BaseCode>,L21:C10-L21:C11,chl=<D>,pl=None]"
- "[@196,196:197=\"r\",<BaseCode>,L21:C11-L21:C12,chl=<D>,pl=None]"
- "[@197,197:198=\"2\",<BaseCode>,L21:C12-L21:C13,chl=<D>,pl=None]"
- "[@198,198:199=\".\",<BaseCode>,L21:C13-L21:C14,chl=<D>,pl=None]"
- "[@199,199:200=\"\\n\",<WS>,L21:C14-L21:C15,chl=<H>,pl=None]"
- "[@200,200:201=\"&\",<BaseCode>,L22:C0-L22:C1,chl=<D>,pl=None]"
- "[@201,201:202=\"&\",<BaseCode>,L22:C1-L22:C2,chl=<D>,pl=None]"
- "[@202,202:203=\"&\",<BaseCode>,L22:C2-L22:C3,chl=<D>,pl=None]"
- "[@203,203:204=\"m\",<BaseCode>,L22:C3-L22:C4,chl=<D>,pl=None]"
- "[@204,204:205=\"v\",<BaseCode>,L22:C4-L22:C5,chl=<D>,pl=None]"
- "[@205,205:206=\"a\",<BaseCode>,L22:C5-L22:C6,chl=<D>,pl=None]"
- "[@206,206:207=\"r\",<BaseCode>,L22:C6-L22:C7,chl=<D>,pl=None]"
- "[@207,207:208=\"&\",<BaseCode>,L22:C7-L22:C8,chl=<D>,pl=None]"
- "[@208,208:209=\"m\",<BaseCode>,L22:C8-L22:C9,chl=<D>,pl=None]"
- "[@209,209:210=\"v\",<BaseCode>,L22:C9-L22:C10,chl=<D>,pl=None]"
- "[@210,210:211=\"a\",<BaseCode>,L22:C10-L22:C11,chl=<D>,pl=None]"
- "[@211,211:212=\"r\",<BaseCode>,L22:C11-L22:C12,chl=<D>,pl=None]"
- "[@212,212:213=\"2\",<BaseCode>,L22:C12-L22:C13,chl=<D>,pl=None]"
- "[@213,213:214=\"\\n\",<WS>,L22:C13-L22:C14,chl=<H>,pl=None]"
- "[@214,214:215=\"&\",<BaseCode>,L23:C0-L23:C1,chl=<D>,pl=None]"
- "[@215,215:216=\"&\",<BaseCode>,L23:C1-L23:C2,chl=<D>,pl=None]"
- "[@216,216:217=\"&\",<BaseCode>,L23:C2-L23:C3,chl=<D>,pl=None]"
- "[@217,217:218=\"m\",<BaseCode>,L23:C3-L23:C4,chl=<D>,pl=None]"
- "[@218,218:219=\"v\",<BaseCode>,L23:C4-L23:C5,chl=<D>,pl=None]"
- "[@219,219:220=\"a\",<BaseCode>,L23:C5-L23:C6,chl=<D>,pl=None]"
- "[@220,220:221=\"r\",<BaseCode>,L23:C6-L23:C7,chl=<D>,pl=None]"
- "[@221,221:222=\"&\",<BaseCode>,L23:C7-L23:C8,chl=<D>,pl=None]"
- "[@222,222:223=\"m\",<BaseCode>,L23:C8-L23:C9,chl=<D>,pl=None]"
- "[@223,223:224=\"v\",<BaseCode>,L23:C9-L23:C10,chl=<D>,pl=None]"
- "[@224,224:225=\"a\",<BaseCode>,L23:C10-L23:C11,chl=<D>,pl=None]"
- "[@225,225:226=\"r\",<BaseCode>,L23:C11-L23:C12,chl=<D>,pl=None]"
- "[@226,226:227=\"2\",<BaseCode>,L23:C12-L23:C13,chl=<D>,pl=None]"
- "[@227,227:228=\".\",<BaseCode>,L23:C13-L23:C14,chl=<D>,pl=None]"
- "[@228,228:229=\"\\n\",<WS>,L23:C14-L23:C15,chl=<H>,pl=None]"
- "[@229,229:230=\"&\",<BaseCode>,L24:C0-L24:C1,chl=<D>,pl=None]"
- "[@230,230:231=\"&\",<BaseCode>,L24:C1-L24:C2,chl=<D>,pl=None]"
- "[@231,231:232=\"&\",<BaseCode>,L24:C2-L24:C3,chl=<D>,pl=None]"
- "[@232,232:233=\"m\",<BaseCode>,L24:C3-L24:C4,chl=<D>,pl=None]"
- "[@233,233:234=\"v\",<BaseCode>,L24:C4-L24:C5,chl=<D>,pl=None]"
- "[@234,234:235=\"a\",<BaseCode>,L24:C5-L24:C6,chl=<D>,pl=None]"
- "[@235,235:236=\"r\",<BaseCode>,L24:C6-L24:C7,chl=<D>,pl=None]"
- "[@236,236:237=\".\",<BaseCode>,L24:C7-L24:C8,chl=<D>,pl=None]"
- "[@237,237:238=\"&\",<BaseCode>,L24:C8-L24:C9,chl=<D>,pl=None]"
- "[@238,238:239=\"m\",<BaseCode>,L24:C9-L24:C10,chl=<D>,pl=None]"
- "[@239,239:240=\"v\",<BaseCode>,L24:C10-L24:C11,chl=<D>,pl=None]"
- "[@240,240:241=\"a\",<BaseCode>,L24:C11-L24:C12,chl=<D>,pl=None]"
- "[@241,241:242=\"r\",<BaseCode>,L24:C12-L24:C13,chl=<D>,pl=None]"
- "[@242,242:243=\"2\",<BaseCode>,L24:C13-L24:C14,chl=<D>,pl=None]"
- "[@243,243:244=\"\\n\",<WS>,L24:C14-L24:C15,chl=<H>,pl=None]"
- "[@244,244:245=\"&\",<BaseCode>,L25:C0-L25:C1,chl=<D>,pl=None]"
- "[@245,245:246=\"&\",<BaseCode>,L25:C1-L25:C2,chl=<D>,pl=None]"
- "[@246,246:247=\"&\",<BaseCode>,L25:C2-L25:C3,chl=<D>,pl=None]"
- "[@247,247:248=\"m\",<BaseCode>,L25:C3-L25:C4,chl=<D>,pl=None]"
- "[@248,248:249=\"v\",<BaseCode>,L25:C4-L25:C5,chl=<D>,pl=None]"
- "[@249,249:250=\"a\",<BaseCode>,L25:C5-L25:C6,chl=<D>,pl=None]"
- "[@250,250:251=\"r\",<BaseCode>,L25:C6-L25:C7,chl=<D>,pl=None]"
- "[@251,251:252=\".\",<BaseCode>,L25:C7-L25:C8,chl=<D>,pl=None]"
- "[@252,252:253=\"&\",<BaseCode>,L25:C8-L25:C9,chl=<D>,pl=None]"
- "[@253,253:254=\"m\",<BaseCode>,L25:C9-L25:C10,chl=<D>,pl=None]"
- "[@254,254:255=\"v\",<BaseCode>,L25:C10-L25:C11,chl=<D>,pl=None]"
- "[@255,255:256=\"a\",<BaseCode>,L25:C11-L25:C12,chl=<D>,pl=None]"
- "[@256,256:257=\"r\",<BaseCode>,L25:C12-L25:C13,chl=<D>,pl=None]"
- "[@257,257:258=\"2\",<BaseCode>,L25:C13-L25:C14,chl=<D>,pl=None]"
- "[@258,258:259=\".\",<BaseCode>,L25:C14-L25:C15,chl=<D>,pl=None]"
- "[@259,259:260=\"\\n\",<WS>,L25:C15-L25:C16,chl=<H>,pl=None]"
- "[@260,260:260=\"<no text>\",<EOF>,L25:C16-L25:C16,chl=<D>,pl=None]"
